designer/server/src/main/scala/pl/touk/nussknacker/ui/component/ComponentsUsageHelper.scala

```diff
@@ -15,21 +16,32 @@ object ComponentsUsageHelper {
   }
 
   def computeComponentsUsage(componentIdProvider: ComponentIdProvider,
-                             processesDetails: List[BaseProcessDetails[ScenarioComponentsUsages]]): Map[ComponentId, List[(BaseProcessDetails[Unit], List[NodeId])]] = {
-
-    def toComponentIdUsages(processDetails: BaseProcessDetails[ScenarioComponentsUsages]): List[(ComponentId, (BaseProcessDetails[Unit], List[NodeId]))] = {
-      val componentsUsages: Map[ComponentIdParts, List[NodeId]] = processDetails.json.value
-      componentsUsages.toList.map { case (ComponentIdParts(componentName, componentType), nodeIds) =>
-        val componentId = componentIdProvider.createComponentId(processDetails.processingType, componentName, componentType)
-        componentId -> (processDetails.mapProcess(_ => ()), nodeIds)
-      }
+                             processesDetails: List[BaseProcessDetails[ScenarioComponentsUsages]]): Map[ComponentId, List[(BaseProcessDetails[_], List[NodeId])]] = {
+    def flattenUsages(processesDetails: List[BaseProcessDetails[ScenarioComponentsUsages]]) = for {
+      processDetails <- processesDetails
+      componentIdNodesPair <- processDetails.json.value.toList
+      (ComponentIdParts(componentName, componentType), nodeIds) = componentIdNodesPair
+      componentId = componentIdProvider.createComponentId(processDetails.processingType, componentName, componentType)
+      nodeId <- nodeIds
+    } yield ScenarioComponentsUsage(componentId, componentType, componentName, processDetails, nodeId)
+
+    val scenariosComponentUsagesFlatten = flattenUsages(processesDetails.filter(_.isSubprocess == false))
+    val fragmentsComponentUsagesFlattenMap = flattenUsages(processesDetails.filter(_.isSubprocess == true))
+      .groupBy(_.processDetails.name).mapValuesNow(_.map(u => u.copy(nodeId = s"<<fragment>> ${u.nodeId}")))
```

Comment:
Can you extract this prefix to const?

Reply:
It'll not live long enough (see my newest comment)